farming/logics/traits/master_chef/farming.rs

```diff
@@ -0,0 +1,482 @@
+
+#[openbrush::trait_definition]
+pub trait Farming: Storage<Data> + Storage<ownable::Data> + FarmingGetters + FarmingEvents {
+    #[ink(message)]
+    #[modifiers(only_owner)]
+    fn add(
+        &mut self,
+        alloc_point: u32,
+        lp_token: AccountId,
+        rewarder: Option<AccountId>,
+    ) -> Result<(), FarmingError> {
+        self._check_pool_duplicate(lp_token)?;
+        self._update_all_pools()?;
+
+        self.data::<Data>().total_alloc_point = self
+            .data::<Data>()
+            .total_alloc_point
+            .checked_add(alloc_point)
+            .ok_or(FarmingError::AddOverflow2)?;
+        self.data::<Data>().lp_tokens.push(lp_token);
+        let pool_length = self.pool_length();
+
+        if let Some(rewarder_address) = rewarder {
+            self.data::<Data>()
+                .rewarders
+                .insert(&pool_length, &rewarder_address);
+        }
+        self.data::<Data>().pool_info.insert(
+            &pool_length,
+            &Pool {
+                acc_arsw_per_share: 0,
+                last_reward_block: Self::env().block_number(),
+                alloc_point,
+            },
+        );
+        self.data::<Data>().pool_info_length = pool_length
+            .checked_add(1)
+            .ok_or(FarmingError::AddOverflow2)?;
+
+        self._emit_log_pool_addition_event(pool_length, alloc_point, lp_token, rewarder);
+        Ok(())
+    }
+
+    #[ink(message)]
+    #[modifiers(only_owner)]
+    fn set(
+        &mut self,
+        pool_id: u32,
+        alloc_point: u32,
+        rewarder: Option<AccountId>,
+        overwrite: bool,
+    ) -> Result<(), FarmingError> {
+        let pool_info = self
+            .get_pool_infos(pool_id)
+            .ok_or(FarmingError::PoolNotFound5)?;
+        self._update_all_pools()?;
+        self.data::<Data>().total_alloc_point = self
+            .data::<Data>()
+            .total_alloc_point
+            .checked_sub(pool_info.alloc_point)
+            .ok_or(FarmingError::SubUnderflow7)?
+            .checked_add(alloc_point)
+            .ok_or(FarmingError::AddOverflow9)?;
+
+        self.data::<Data>().pool_info.insert(
+            &pool_id,
+            &Pool {
+                alloc_point,
+                ..pool_info
+            },
+        );
+        let mut rewarder = rewarder;
+        if overwrite {
+            match rewarder {
+                Some(rewarder_address) => {
+                    self.data::<Data>()
+                        .rewarders
+                        .insert(&pool_id, &rewarder_address)
+                }
+                None => self.data::<Data>().rewarders.remove(&pool_id),
+            }
+        } else {
+            rewarder = self.get_rewarder(pool_id);
+        }
+        self._emit_log_set_pool_event(pool_id, alloc_point, rewarder, overwrite);
+        Ok(())
+    }
+
+    #[ink(message)]
+    fn pending_arsw(&mut self, pool_id: u32, user: AccountId) -> Result<Balance, FarmingError> {
```

Comment:
```suggestion
    fn pending_arsw(&self, pool_id: u32, user: AccountId) -> Result<Balance, FarmingError> {
```
pending_arsw is a view function.